File: document/src/vespa/document/datatype/arraydatatype.cpp
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include "arraydatatype.h"
#include <vespa/document/fieldvalue/arrayfieldvalue.h>
#include <vespa/vespalib/util/exceptions.h>
#include <ostream>

namespace document {

ArrayDataType::ArrayDataType(const DataType &nestedType, int32_t id)
    : CollectionDataType("Array<" + nestedType.getName() + ">", nestedType, id)
{
}

ArrayDataType::ArrayDataType(const DataType& nestedType)
    : CollectionDataType("Array<" + nestedType.getName() + ">", nestedType)
{
}

FieldValue::UP
ArrayDataType::createFieldValue() const
{
    return std::make_unique<ArrayFieldValue>(*this);
}

void
ArrayDataType::print(std::ostream& out, bool verbose,
                     const std::string& indent) const
{
    out << "ArrayDataType(\n" << indent << "    ";
    getNestedType().print(out, verbose, indent + "    ");
    out << ", id " << getId() << ")";
}

bool
ArrayDataType::equals(const DataType& other) const noexcept
{
    if (this == &other) return true;
    return CollectionDataType::equals(other) && other.isArray();
}

void
ArrayDataType::onBuildFieldPath(FieldPath & path, vespalib::stringref remainFieldName) const
{
    if (remainFieldName[0] == '[') {
        size_t endPos = remainFieldName.find(']');
        if (endPos == vespalib::stringref::npos) {
            throw vespalib::IllegalArgumentException("Array subscript must be closed with ]");
        } else {
            int pos = endPos + 1;
            if (remainFieldName[pos] == '.') {
                pos++;
            }

            getNestedType().buildFieldPath(path, remainFieldName.substr(pos));

            if (remainFieldName[1] == '$') {
                path.insert(path.begin(), std::make_unique<FieldPathEntry>(getNestedType(), remainFieldName.substr(2, endPos - 2)));
            } else {
                // FIXME C++17 range-safe from_chars() instead of atoi()
                path.insert(path.begin(), std::make_unique<FieldPathEntry>(getNestedType(), atoi(remainFieldName.substr(1, endPos - 1).data())));
            }
        }
    } else {
        getNestedType().buildFieldPath(path, remainFieldName);
    }
}

} // document

File: document/src/main/java/com/yahoo/document/ArrayDataType.java
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.document;

import com.yahoo.document.datatypes.Array;
import com.yahoo.vespa.objects.Ids;

import java.util.ArrayList;
import java.util.List;

/**
 * @author Einar M R Rosenvinge
 */
public class ArrayDataType extends CollectionDataType {

    // The global class identifier shared with C++.
    public static int classId = registerClass(Ids.document + 54, ArrayDataType.class);

    public ArrayDataType(DataType nestedType) {
        super("Array<"+nestedType.getName()+">", 0, nestedType);
        setId(getName().toLowerCase().hashCode());
    }

    public ArrayDataType(DataType nestedType, int code) {
        super("Array<"+nestedType.getName()+">", code, nestedType);
    }

    public ArrayDataType clone() {
        return (ArrayDataType) super.clone();
    }

    public Array createFieldValue() {
        return new Array(this);
    }

    @Override
    public Class getValueClass() {
        return Array.class;
    }

    @Override
    public FieldPath buildFieldPath(String remainFieldName)
    {
        if (remainFieldName.length() > 0 && remainFieldName.charAt(0) == '[') {
            int endPos = remainFieldName.indexOf(']');
            if (endPos == -1) {
                throw new IllegalArgumentException("Array subscript must be closed with ]");
            } else {
                FieldPath path = getNestedType().buildFieldPath(skipDotInString(remainFieldName, endPos));
                List<FieldPathEntry> tmpPath = new ArrayList<FieldPathEntry>(path.getList());
                if (remainFieldName.charAt(1) == '$') {
                    tmpPath.add(0, FieldPathEntry.newVariableLookupEntry(remainFieldName.substring(2, endPos), getNestedType()));
                } else {
                    tmpPath.add(0, FieldPathEntry.newArrayLookupEntry(Integer.parseInt(remainFieldName.substring(1, endPos)), getNestedType()));
                }

                return new FieldPath(tmpPath);
            }
        }

        return getNestedType().buildFieldPath(remainFieldName);
    }

}
